sponsorship_compassion/models/contracts.py

```diff
@@ -560,8 +545,11 @@ def write(self, vals):
             old_partners.update_number_sponsorships()
 
         if "correspondent_id" in vals:
-            self.mapped("correspondent_id").update_number_sponsorships()
-            old_correspondents.update_number_sponsorships()
+            new_correspondents = vals["correspondent_id"]
+            new_correspondents = self.env["res.partner"].browse(new_correspondents)
+            for record in self:
+                record._add_correspondent(new_correspondents)
+                record.correspondent_id.update_number_sponsorships()
```

Comment:
You should move this inside `_add_correspondent` method, just before the commit.